crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+        let orthonormal_eigenvectors = QR::new(eigh.eigenvectors).q();
+        (eigvals, orthonormal_eigenvectors)
+    } else {
+        let schur = nalgebra::linalg::Schur::try_new(um0um1, 1e-12, 100000).unwrap();
+        let (vmat, evals) = schur.unpack();
+        let eigvals = evals.diagonal();
+        (eigvals, vmat)
+    };
+    let d_values: DVector<Complex64> = eigvals.map(|x| x.sqrt());
+    let d_mat: DMatrix<Complex64> = DMatrix::from_diagonal(&d_values);
+    let wmat = &d_mat * vmat.adjoint() * &um1;
+
+    debug_assert!(_demultiplex_verify(&um0, &um1, &vmat, &wmat, &d_mat));
+
+    let out_qubits = (0..num_qubits)
+        .map(|_| ShareableQubit::new_anonymous())
+        .collect::<Vec<_>>();
+    let mut out = CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.))?;
+
+    // left gate. In this case we decompose wmat.
+    // Otherwise, it is combined with the B matrix.
+    match vw_type {
+        VWType::OnlyW | VWType::All => {
+            let left_circuit = qsd_inner(&wmat, Some(opt_a1), Some(opt_a2), None, None, depth + 1)?;
+            append(&mut out, left_circuit, &layout[..num_qubits - 1])?;
+        }
+        VWType::OnlyV => (),
+    }
+
+    // multiplexed Rz gate
+    // If opt_a1 = ``True``, then we reduce 2 ``cx`` gates per call.
+    let mut angles = d_values
+        .iter()
+        .map(|x| 2. * x.conj().arg())
+        .collect::<Vec<_>>();
+    let ucrz = match (&vw_type, opt_a1) {
+        (VWType::OnlyW, true) => get_ucrz(num_qubits, &mut angles, false)?,
+        (VWType::OnlyV, true) => get_ucrz(num_qubits, &mut angles, false)?.reverse()?,
+        _ => get_ucrz(num_qubits, &mut angles, true)?,
+    };
+    // let multiplexed_rz = decompose_uc_rotation(&mut angles, StandardGate::RZ)?;
+    append(
+        &mut out,
+        ucrz,
+        &[
+            std::slice::from_ref(layout.last().unwrap()),
+            &layout[..num_qubits - 1],
+        ]
+        .concat(),
+    )?;
+    // right gate. In this case we decompose vmat.
+    // Otherwise, it is combined with the B matrix.
+    match vw_type {
+        VWType::OnlyV | VWType::All => {
+            let right_circuit =
+                qsd_inner(&vmat, Some(opt_a1), Some(opt_a2), None, None, depth + 1)?;
+            append(&mut out, right_circuit, &layout[..num_qubits - 1])?;
+        }
+        VWType::OnlyW => (),
+    }
+    Ok((out, vmat, wmat))
+}
+
+#[allow(clippy::toplevel_ref_arg)]
+fn _demultiplex_verify(
+    um0: &DMatrix<Complex64>,
+    um1: &DMatrix<Complex64>,
+    vmat: &DMatrix<Complex64>,
+    wmat: &DMatrix<Complex64>,
+    dmat: &DMatrix<Complex64>,
+) -> bool {
+    let u_block = stack![um0, 0; 0, um1];
+    let v_block = stack![vmat, 0; 0, vmat];
+    let w_block = stack![wmat, 0; 0, wmat];
+    let d_inv = dmat.adjoint();
+    let d_block = stack![dmat, 0; 0, d_inv];
+
+    let u_check = v_block * d_block * w_block;
+
+    const EPS: f64 = 1e-7;
+    abs_diff_eq!(u_block, u_check, epsilon = EPS)
```

Comment:
Defining a constant like this right before it's only use is a bit odd. It doesn't hurt anything but it's just using two lines for what is identical to:

```suggestion
    abs_diff_eq!(u_block, u_check, epsilon = 1e-7)
```

If we were reusing `EPS` as 1e-7 elsewhere in the file defining a shared `const` value here would be a good choice. Although this shadows an existing `EPS` constant value of 1e-10 which is also not ideal.